Structure

jholbrook64 opened the pull request:
This commit creates four functions to easily run a workflow in R that takes data from the web and generates a PNG, .txt. and .csv output in three distinct folders.

Functions are:

getCSV(out_folder) - takes argument for path to save data to a file in.

eval_data(in_csv) - takes input data and creates a data frame with required column names.

createPlot(evalData, plotOutput) - takes a data frame and an output file and generate a plot in that file.

dataRender(eval_dtaa, textOutput) - takes a data frame and an output file and writes a descriptive piece of text about the data contained within the data frame.

to run the code, add each of the four functions to the
environment. Then run all the pieces of code that create the desired output directories and then run the code that calls these functions.

jholbrook64 committed:
Merge pull request #4 from jholbrook64/main

update structure

jholbrook64 committed:
These comits are made to structure so that strucutre can be merged with
main, I am going to open a new pull request so I can pull request in the
right order with this commit, going    main  <-    structure

limnoliver commented:
Hey Jack - can you leave a comment with instructions for how you would execute your functions (per the instructions [here)](https://github.com/jholbrook64/ds-pipelines-targets-1/pull/4)

limnoliver commented on 1_fetch/src/my_happy_script.R:
```diff
@@ -1,96 +1,133 @@
+# Get the data from Science Base
+getCSV <- function(){
+  mendota_file <- file.path(project_output_dir, 'model_RMSEs.csv')
+  item_file_download('5d925066e4b0c4f70d0d0599', names = 'me_RMSE.csv', destinations = mendota_file, overwrite_file = TRUE)
```
This is "fetching" data from Sciencebase. Now that you have a "1_fetch" step, the output of this function should write to "1_fetch/out"

limnoliver commented on 1_fetch/src/my_happy_script.R:
```diff
@@ -1,96 +1,133 @@
-mendota_file <- file.path(project_output_dir, 'model_RMSEs.csv')
-item_file_download('5d925066e4b0c4f70d0d0599', names = 'me_RMSE.csv', destinations = mendota_file, overwrite_file = TRUE)
+# Prepare the data for plotting
+eval_data <- function(getCSV){
```
This function should be accepting the output of the getCSV function. While you could name this argument getCSV, I think this is confusing. Perhaps the argument should be named something clearer, like "lake_csv" or "in_csv".

limnoliver commented on 1_fetch/src/my_happy_script.R:
```diff
@@ -1,96 +1,133 @@
-item_file_download('5d925066e4b0c4f70d0d0599', names = 'me_RMSE.csv', destinations = mendota_file, overwrite_file = TRUE)
+# Prepare the data for plotting
+eval_data <- function(getCSV){
+  eval_data <- readr::read_csv(getCSV, col_types = 'iccd') %>% # switch mendota_file to function call
```
While you could have this object be named `eval_data` within the function, this is confusing since the function has the same name. I would name this something like `out_data` or `munged_data`

limnoliver commented on 1_fetch/src/my_happy_script.R:
```diff
@@ -1,96 +1,133 @@
-
-# Create a plot
+# Create a plot, this function won't return anything
+#createPlot <- function(cleanedData){
```
Should turn this into a function like you have here, but is commented out.

limnoliver commented on 1_fetch/src/my_happy_script.R:
```diff
@@ -1,96 +1,133 @@
+# Create a plot, this function won't return anything
+#createPlot <- function(cleanedData){
+
 png(file = file.path(project_output_dir, 'figure_1.png'), width = 8, height = 10, res = 200, units = 'in')
```
If you do stick with a `3_visualize` step, this should go in `3_visualize/out`.

limnoliver commented on 1_fetch/src/my_happy_script.R:
```diff
@@ -1,96 +1,133 @@
 
-
 # Save the processed data
 readr::write_csv(eval_data, file = file.path(project_output_dir, 'model_summary_results.csv'))
```
This should go in a separate function that writes or returns the data object. I think it makes sense to have a separate data munge and plotting function.

limnoliver replied:
And sorry I wasn't clearer here. I think you can return an object for the eval data (as you have in function `eval_data`  and you _don't_ need to return it as a csv as you're doing here. I think the `eval_data` function does what you need, and you can ingest that object in other functions that rely on it. That is, I don't think this extra csv is required.

limnoliver commented on 1_fetch/src/my_happy_script.R:
```diff
@@ -1,96 +1,133 @@
   The PGDL prediction accuracy was more robust compared to PB when only two profiles were provided for training ({{pgdl_2mean}} and {{pb_2mean}}°C, respectively). '
 
-whisker.render(template_1 %>% str_remove_all('\n') %>% str_replace_all('  ', ' '), render_data ) %>% cat(file = file.path(project_output_dir, 'model_diagnostic_text.txt'))
+whisker.render(template_1 %>% str_remove_all('\n') %>% str_replace_all('  ', ' '), render_data ) %>% cat(file = file.path(project_output_dir, 'model_diagnostic_text.txt'))
```
Need to close this function out with a `}` at the end.

jholbrook64 committed:
Commited changes to wrap all code that handles the mendota file data
into functions. Functions include:
				getCSV()
				eval_data()
				createPlot()
				out_csv()
				dataRender()
				whiskerRender()
The next push will be changes to make the functions output files into
the 1_fetch/out/my_dir/ directory.

limnoliver commented on 1_fetch/src/my_happy_script.R:
```diff
@@ -20,47 +20,50 @@ library(stringr)
 
-# Get the data from Science Base
+# Get the data from Science Base    # good
 getCSV <- function(){
```
I recommend including an argument called `out_folder` or something similar, which is set to `1_fetch/out` which then allows `menota_file` to be set as `1_fetch/out/model_RMSEs.csv` in line 29 if you change it to 

`mendota_file <- file.path(out_folder, 'model_RMSE.csv')`

limnoliver commented on 1_fetch/src/my_happy_script.R:
```diff
@@ -11,7 +11,7 @@
 ## ---------------------------
 
-setwd("C:/Users/jholbrook/Documents/Rprojects/targets-1")    ## Jacks's working directory (PC)
+setwd("C:/Users/jholbrook/Documents/Rprojects/targets-1/")    ## Jacks's working directory (PC)
```
When you're in an R Project everthing will be relative to `C:/Users/jholbrook/Documents/Rprojects/targets-1` or whatever your project path is. It's best practice to _not_ use `setwd` because a collaborator would have to change this path to run this on their own machine.

limnoliver commented on 1_fetch/src/my_happy_script.R:
```diff
@@ -107,27 +105,65 @@ text(2.3, 1.1, 'Process-Based', pos = 4, cex = 1.1)
+return(render_data)
+     }
+
+whiskerRender <- function(dataRender){
```
My personal opinion, but I think `dataRender` and `whiskerRender` could be a single function instead of two. Then you're not reading in the eval_data twice.

jholbrook64 committed:
Commited changes:
		Reduced total number of functions to four functions:

		getCSV() takes 1 parameter of the output directory
		path
		eval_data() reads the csv file from the get function and
		returns a dataframe
		createPlot() reads in the dataframe and generate and
		exports a plot to the target directory
		dataRender() takes the dataframe and saves a text file
		to the output directory.

limnoliver commented on 1_fetch/src/my_happy_script.R:
```diff
@@ -20,16 +19,18 @@ library(stringr)
-getCSV <- function(){
-  mendota_file <- file.path(project_output_dir, 'model_RMSEs.csv')
+#include a parameter for the path to the output folder from tHe project directory:
+out_folder <-  "1_fetch/out/my_dir"
```
This can just be `1_fetch/out`

limnoliver replied:
Also, you can move this down to the "instructions" section that shows you how to use the functions.

limnoliver commented on 1_fetch/src/my_happy_script.R:
```diff
@@ -20,16 +19,18 @@ library(stringr)
+# create new output directory
 project_output_dir <- 'my_dir'   # 1_fetch/out/my_dir
-
 dir.create(project_output_dir)
```
maybe move this to the instructions section as well since it's not part of the functinos.

limnoliver commented on 1_fetch/src/my_happy_script.R:
```diff
@@ -127,43 +124,33 @@ render_data <- list(pgdl_980mean = filter(eval_data, model_type == 'pgdl', exper
-# not sure which of these formats to do, the top one throws an error, this error can be resolved 
-# with passing the inner as a function rather than the function's name. Passing the function however doesn't output the file.
+#create object for evaluated data:
+evalData <- eval_data(getCSV(out_folder))
```
Lines 142 and 145 will download the data twice. You could do the following to avoid that:

```
file_out <- getCSV(out_folder)
evalData <- eval_data(file_out)
```

limnoliver commented on 1_fetch/src/my_happy_script.R:
```diff
@@ -51,7 +52,7 @@ eval_data <- function(in_csv){
 
 # Create a plot, this function won't return anything
-createPlot <- function(eval_data){
+createPlot <- function(evalData){
```
this could have an `out_folder` argument as well that you can set to `3_visualize/out` or whatever you want to call the 3 step.

limnoliver commented on 1_fetch/src/my_happy_script.R:
```diff
@@ -127,43 +124,33 @@ render_data <- list(pgdl_980mean = filter(eval_data, model_type == 'pgdl', exper
-# this shows that the funciton can RETURN the values, but won't write them to the file in the directory
-outPut <- out_csv(eval_data(getCSV()))
-outPut   
+thePlot <- createPlot(evalData)   
```
This you can just run and not save as an object. It will write the file, but you don't need to return anything from the function for now (later, you might need to return a file name for the `targets` pipeline setup). So, this line should just be:

`createPlot(evalData, out_folder = `3_visualize/out`)` after you modify the arguments as suggested above.

limnoliver commented on 1_fetch/src/my_happy_script.R:
```diff
@@ -127,43 +124,33 @@ render_data <- list(pgdl_980mean = filter(eval_data, model_type == 'pgdl', exper
+thePlot                                     
 
-renderList <- dataRender(eval_data(getCSV()))
+renderList <- dataRender(evalData)
```
Same thing here: include an `out_folder` argument, and then down in your examples it should look like:

`dataRender(evalData, out_folder = '2_process/out')`

jholbrook64 committed:
Commit fully working functionalized code.
	Functions:
	getCSV(out_folder) - takes argument for path to save data to
	eval_data(in_csv) - takes input data and creates a dataframe with
	required column names
	createPlot(evalData, plotOutput) - takes a dataframe and an
	output file and generate a plot in that file
	dataRender(eval_dtaa, textOutput) - takes a dataframe and an
	output file and writes a descriptive piece of text about the
	data contained within the dataframe

	to run the code, add each of the four functions to the
	environment. Then run all the pieces of code that create the
	desired output directories and then run the code that calls
	these functions.

jholbrook64 commented on 1_fetch/src/my_happy_script.R:
```diff
@@ -1,83 +1,108 @@
+## ---------------------------
```
This commit creates four functions to easily run a workflow in R that  takes data from the web and generates a PNG, .txt. and .csv output in three distinct folders. 

Functions are:

getCSV(out_folder) - takes argument for path to save data to a file in.

eval_data(in_csv) - takes input data and creates a data frame with required column names.
	
createPlot(evalData, plotOutput) - takes a data frame and an output file and generate a plot in that file.

dataRender(eval_dtaa, textOutput) - takes a data frame and an output file and writes a descriptive piece of text about the data contained within the data frame.

to run the code, add each of the four functions to the
environment. Then run all the pieces of code that create the desired output directories and then run the code that calls these functions.

limnoliver commented on 1_fetch/src/my_happy_script.R:
```diff
@@ -87,10 +112,45 @@ render_data <- list(pgdl_980mean = filter(eval_data, model_type == 'pgdl', exper
+dir.create(visualize_output_dir)
+dir.create(process_output_dir)
+dir.create(Visualize)
+dir.create(visualize_src)
```
If you were using github and collaborating with someone on this code, you wouldn't need to create a directory each time because in many cases they would already exist. For example, any `src` directory that contains committed code would already exist. In some cases, you might not be git tracking data (like the csv you downloaded), so you may want to `dir.create` the out folders. I think you could simplify the above code chunk quite a bit. You're not using `Visualize` anywhere, so can probably delete line 131. You can delete 133 and 138 because those filepaths will exist once you commit the code to that place.

jholbrook64 committed:
This commit creates four functions to easily run a workflow in R that
takes data from the web and generates a PNG, .txt. and .csv output in
three distinct folders.

Functions are:

getCSV(out_folder) - takes argument for path to save data to a file in.

eval_data(in_csv) - takes input data and creates a data frame with
required column names.

createPlot(evalData, plotOutput) - takes a data frame and an output file
and generate a plot in that file.

dataRender(eval_dtaa, textOutput) - takes a data frame and an output
file and writes a descriptive piece of text about the data contained
within the data frame.

to run the code, add each of the four functions to the
environment. Then run all the pieces of code that create the desired
output directories and then run the code that calls these functions.

limnoliver commented on 1_fetch/src/my_happy_script.R:
```diff
@@ -87,10 +110,44 @@ render_data <- list(pgdl_980mean = filter(eval_data, model_type == 'pgdl', exper
+
+# create directories from associated paths: these are kept because functions will throw errors if these don't exist 
+dir.create(process_output_dir)
+dir.create(Visualize)
```
this should be

`dir.create(visualize_output_dir)`

limnoliver reviewed (changes requested):
One final change, and then you can merge. Great work!

jholbrook64 committed:
This commit removes uneeded code to create  directories.

	Functions are:

	getCSV(out_folder) - takes argument for path to save data to a
	file in.

	eval_data(in_csv) - takes input data and creates a data frame
	with required column names.

	createPlot(evalData, plotOutput) - takes a data frame and an
	output file and generate a plot in that file.

	dataRender(eval_dtaa, textOutput) - takes a data frame and an
	output file and writes a descriptive piece of text about the
	data contained within the data frame.

	to run the code, add each of the four functions to the
	environment. Then run all the pieces of code that create the
	desired output directories and then run the code that calls
	these functions.

jholbrook64 commented:
> 
> 
> Great comments @jholbrook64! ✨
> 
> You could consider [GNU make](https://www.gnu.org/software/make/) to be a great grandparent of the packages we referred to early in this lesson (`remake`, `scipiper`, `drake`, and `targets`). Will Landau, the lead developer of `targets`, has added a lot of useful features to dependency management systems in R, and has a great way of summarizing why we put energy into using these tools: **"Skip the work you don't need"**
> 
> We'd like you to next check out a short part of Will's video on `targets`
> 
> [![reproducible workflows with R targets](https://user-images.githubusercontent.com/2349007/81425244-cea12300-911c-11ea-8555-0c5a61fbee21.png)](https://www.youtube.com/watch?v=Gqn7Xn4d5NI)
> ### 📺 Activity: watch video on targets from at least 7:20 to 11:05 (you are welcome to watch the full main talk from 7:20 to 7:40)
> 
> Use a github comment on this issue to let us know what contrasts you identified between solutions in `make` and what is offered in R-specific tools, like `targets`. Please use less than 300 words. Then assign your onboarding cohort team member this issue to read what you wrote and respond with any questions or comments.
> ### When you are satisfied with the discussion, you can close this issue and I'll direct you to the next one.

The contrasts between GNU Make and R Targets
GNU make is language independent since it relies on sets of instructions called make lists that communicate to the compiler, targets works specifically in the R environment and abstracts files as R objects to speak to the compiler.
Originally, GNU Make identified functions in files as targets, targets have prerequisites and dependencies that the code, or “recipe” needs to in order to make the target. The R package targets organizes its objects in a similar order which makes it a “Make-like” pipeline tool. The goal of R targets is to create an environment like a package where prerequisites and dependencies are handled so as to preserve the data being fed into the modelling and visualizing portions of a workflow. 
In contrast, GNU Make’s most common use case is to help packages update themselves more efficiently.  Make can also be used as a pipeline tool. This can be done where the functions to model and visualize data are separate Make targets and data is passed to the targets. This could work like:
	Cat    data_in  |  Model.R  |  > data_tmp            # cat concatenates and writes files
	Cat    data_tmp  | Visualize.R  | > data_out
In targets, the data pipeline would work similar to this where the modelling and visualizing components would also be in separate objects. Targets is special because it includes a configuration file _Targets.R that loads the packages and functions that define the workflow. Inside the file, a pipeline object is also defined. The pipeline object behaves similarly to the GNU make file where there is one pipeline object defined for each function, just like how there is one target defined for each function in GNU Make.
